### ISSUE TITLE
ERROR: Can not execute `setup.py` since setuptools is not available in the build environment.

### ISSUE THREAD

**Issue by Xiaotian0726** [romesco/hydra-lightning#19]
Error occurs when I install this repo with pip:

![1](https://user-images.githubusercontent.com/49309571/160238025-1a39ca04-d797-4b9f-9bad-63aa8e6a87c5.jpg)

I've tried some methods on internet about this ERROR message, like upgrade or downgrade `setuptools`, but all failed. Could anybody help me with it? Thank you!

**My environment**: 
- OS: Ubuntu 18.04
- Python: 3.7.13 (venv)

**My current pip list**: 
Package                  Version
absl-py                  1.0.0
aiohttp                  3.8.1
aiosignal                1.2.0
antlr4-python3-runtime   4.8
appdirs                  1.4.4
async-timeout            4.0.2
asynctest                0.13.0
attrs                    21.4.0
audioread                2.1.9
autopage                 0.5.0
backcall                 0.2.0
beautifulsoup4           4.10.0
cachetools               5.0.0
certifi                  2021.10.8
cffi                     1.15.0
charset-normalizer       2.0.12
cmaes                    0.8.2
colorlog                 6.6.0
ctcdecode                1.0.3
cycler                   0.11.0
debugpy                  1.6.0
decorator                5.1.1
defusedxml               0.7.1
dnspython                2.2.1
entrypoints              0.4
fonttools                4.31.2
frozenlist               1.3.0
fsspec                   2022.2.0
future                   0.18.2
google-auth              2.6.2
google-auth-oauthlib     0.4.6
google-crc32c            1.3.0
google-resumable-media   2.3.2
googleapis-common-protos 1.56.0
greenlet                 1.1.2
grpcio                   1.44.0
hydra-core               1.1.1
idna                     3.3
importlib-metadata       4.11.3
importlib-resources      5.6.0
iniconfig                1.1.1
ipython                  7.32.0
ipython-genutils         0.2.0
itsdangerous             2.1.2
jedi                     0.18.1
Jinja2                   3.1.1
joblib                   1.1.0
jsonschema               4.4.0
jupyter-client           7.1.2
jupyter-core             4.9.2
jupyterlab-pygments      0.1.2
jupyterlab-widgets       1.1.0
kiwisolver               1.4.0
llvmlite                 0.38.0
Mako                     1.2.0
Markdown                 3.3.6
MarkupSafe               2.1.1
matplotlib               3.5.1
matplotlib-inline        0.1.3
mistune                  0.8.4
multidict                6.0.2
nest-asyncio             1.5.4
numba                    0.55.1
numpy                    1.21.5
oauthlib                 3.2.0
omegaconf                2.1.1
packaging                21.3
pandocfilters            1.5.0
parso                    0.8.3
pbr                      5.8.1
pexpect                  4.8.0
pickleshare              0.7.5
Pillow                   9.0.1
pip                      22.0.4
pluggy                   1.0.0
pooch                    1.6.0
prettytable              3.2.0
prometheus-client        0.13.1
prompt-toolkit           3.0.28
protobuf                 3.19.4
psutil                   5.9.0
ptyprocess               0.7.0
py                       1.11.0
pyasn1                   0.4.8
pyasn1-modules           0.2.8
pycparser                2.21
pyDeprecate              0.3.1
Pygments                 2.11.2
pyparsing                3.0.7
pyperclip                1.8.2
pyrsistent               0.18.1
python-dateutil          2.8.2
python-etcd              0.4.5
python-Levenshtein       0.12.2
pytorch-lightning        1.5.10
PyYAML                   6.0
pyzmq                    22.3.0
QtPy                     2.0.1
requests                 2.27.1
requests-oauthlib        1.3.1
rsa                      4.8
scikit-learn             1.0.2
scipy                    1.7.3
Send2Trash               1.8.0
setuptools               59.5.0
six                      1.16.0
SoundFile                0.10.3.post1
soupsieve                2.3.1
sox                      1.4.1
SQLAlchemy               1.4.32
stevedore                3.5.0
tensorboard              2.8.0
tensorboard-data-server  0.6.1
tensorboard-plugin-wit   1.8.1
terminado                0.13.3
testpath                 0.6.0
threadpoolctl            3.1.0
tomli                    2.0.1
torch                    1.10.0
torchaudio               0.10.0+cu111
torchelastic             0.2.0
torchmetrics             0.7.3
tornado                  6.1
tqdm                     4.63.1
traitlets                5.1.1
typing_extensions        4.1.1
urllib3                  1.26.9
wcwidth                  0.2.5
webencodings             0.5.1
Werkzeug                 2.0.3
wget                     3.2
wheel                    0.37.1
yarl                     1.7.2
zipp                     3.7.0

**Comment by Xiaotian0726** [comment]
This problem has been solved.

References:
https://github.com/pypa/packaging-problems/issues/573

**Comment by drobison00** [comment]
For anyone else who happens to land here looking for a solution on Linux:

If you are building `cpython` from source, double check that `_ctypes` is actually being built and that you have installed `libffi-dev` in the build environment `apt install libffi-dev` or something like `conda install libffi`. The error above can mask the missing _ctypes library, because pip will try to import setuptools which will fail because of missing _ctypes, this causes pip to claim setuptools isn't installed.
